ansible_collections/arista/avd/roles/eos_designs/python_modules/connected_endpoints/utils.py

```diff
@@ -0,0 +1,287 @@
+from __future__ import annotations
+
+import re
+from functools import cached_property
+from hashlib import sha256
+
+from ansible_collections.arista.avd.plugins.filter.convert_dicts import convert_dicts
+from ansible_collections.arista.avd.plugins.filter.esi_management import generate_esi, generate_route_target
+from ansible_collections.arista.avd.plugins.plugin_utils.merge import merge
+from ansible_collections.arista.avd.plugins.plugin_utils.utils import default, get, get_item
+from ansible_collections.arista.avd.roles.eos_designs.python_modules.interface_descriptions import AvdInterfaceDescriptions
+from ansible_collections.arista.avd.roles.eos_designs.python_modules.ip_addressing import AvdIpAddressing
+
+
+class UtilsMixin:
+    """
+    Mixin Class with internal functions.
+    Class should only be used as Mixin to a AvdStructuredConfig class
+    """
+
+    # Set type hints for Attributes of the main class as needed
+    _hostvars: dict
+    _avd_ip_addressing: AvdIpAddressing
+    _avd_interface_descriptions: AvdInterfaceDescriptions
+
+    @cached_property
+    def _filtered_connected_endpoints(self) -> list:
+        """
+        Return list of endpoints defined under one of the keys in "connected_endpoint_keys"
+        which are connected to this switch.
+
+        Adapters are filtered to contain only the ones connected to this switch.
+        """
+        filtered_connected_endpoints = []
+        for connected_endpoints_key in self._connected_endpoints_keys:
+            connected_endpoints = convert_dicts(get(self._hostvars, connected_endpoints_key["key"], default=[]), "name")
+            for connected_endpoint in connected_endpoints:
+                if "adapters" not in connected_endpoint:
+                    continue
+
+                connected_endpoint["adapters"] = [adapter for adapter in connected_endpoint["adapters"] if self._hostname in adapter.get("switches", [])]
+
+                for adapter_index, adapter in enumerate(connected_endpoint["adapters"]):
+                    if "profile" not in adapter:
+                        continue
+
+                    port_profile = get_item(self._merged_port_profiles, "profile", adapter["profile"], default={})
+                    adapter = merge(port_profile, adapter, list_merge="replace", destructive_merge=False)
+                    adapter.pop("profile")
+                    connected_endpoint["adapters"][adapter_index] = adapter
```

Comment:
Wondering if this is the most efficient to first recreate a new list of adapters line 41 by filtering and then reiterating over all of them a second time to handle profile.

Probably best to build the list once and handle everything in one iteration through the list.

```suggestion
                filtered_adapters = []
                for adapter in connected_endpoint["adapters"]:
                    if self._hostname not in adapter.get("switches", [])]:
                        continue
                     
                    if "profile" in adapter:
                        port_profile = get_item(self._merged_port_profiles, "profile", adapter["profile"], default={})
                        adapter = merge(port_profile, adapter, list_merge="replace", destructive_merge=False)
                        adapter.pop("profile")
                    
                    filtered_adapters.append(adapter)
```

Reply:
closing as agreed. Not worth it for small amount of adapters per endpoint.

Reply:
Ended up doing a variant of this any way, because we need to merge profiles before checking if device is part of "switches". to cover the case where "switches" is set on the profile